README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# mpower

<!-- badges: start -->

[![R-CMD-check](https://github.com/phuchonguyen/mpower/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/phuchonguyen/mpower/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

This package allows users to conduct power analysis based on Monte Carlo
simulations in settings in which considerations of the correlations
between predictors are important. It runs power analyses given a data
generative model and an inference model. It can set up a data generative
model that preserves dependence structures among variables given
existing data (continuous, binary, or ordinal) or high-level
descriptions of the associations. Users can generate power curves to
assess the trade-offs between sample size, effect size, and power of a
design.

This vignette presents tutorials and examples focusing on applications
for environmental mixtures studies where predictors tend to be
moderately to highly correlated. It easily interfaces with several
existing and newly developed analysis strategies for assessing
associations between exposures to mixtures and health outcomes. However,
the package is sufficiently general to facilitate power simulations in a
wide variety of settings.

## Installation

And the development version from
[GitHub](https://github.com/phuchonguyen/mpower) with:

``` r
# install.packages("devtools")
devtools::install_github("phuchonguyen/mpower", ref="mixselect")
```

## Tutorial

``` r
library(mpower)
library(dplyr)
library(magrittr)
library(tidyselect)
library(ggplot2)
```

To do power analysis using Monte Carlo simulations, we’ll need:

-   A generative model for the predictors. The predictors can be
    correlated and mixed-scaled. Use `MixtureModel()`.

-   A generative model for the outcome that describes the “true”
    relationships between the predictors and outcome. Use
    `OutcomeModel()`.

-   An inference model. Ideally, this would be the same as the
    generative model for the outcome. However, in practice our model can
    at best approximate the “true” predictor-outcome relationships. Use
    `InferenceModel()`.

-   A “significance” criterion and threshold for the inference model.
    For instance, the t-test in a linear regression has its p-value as
    the “significance” criterion, and a common threshold for statistical
    significance is p-value less than 0.05.

### Example 1: Power curve

#### Generate predictors with C-vine

We can manually specify the joint distribution of the predictors using
univariate marginal distributions and a guess of the correlation matrix.
To create four moderately associated mixed-scaled predictors:

``` r
G <- diag(1, 4)
G[upper.tri(G)] <- 0.6
G[lower.tri(G)] <- 0.6
xmod <- mpower::MixtureModel(method = "cvine", G = G, m = 50, 
    cvine_marginals = list(binary = "qbinom(size=1, prob=0.2)", 
                           count = "qpois(lambda = 1)",
                           ordinal = "qmultinom(probs=c(0.6, 0.2, 0.2))",
                           categorical = "qmultinom(probs=c(0.6, 0.2, 0.2))"),
    cvine_dtypes = list(categorical = "factor"))
mpower::mplot(xmod)
```

<img src="man/figures/README-cvine-xmod-1.png" width="100%" /><img src="man/figures/README-cvine-xmod-2.png" width="100%" /><img src="man/figures/README-cvine-xmod-3.png" width="100%" />

#### Define outcome models

Since we want to create a power curve, we need to supply a list of
outcome models with different “true” effect sizes:

``` r
ymod_list <- list(
  mpower::OutcomeModel(f = "1*binary + 1*I(categorical==1)", sigma = 1, family = "gaussian"),
  mpower::OutcomeModel(f = "0.5*binary + 0.5*I(categorical==1)", sigma = 1, family = "gaussian"),
  mpower::OutcomeModel(f = "0.2*binary + 0.2*I(categorical==1)", sigma = 1, family = "gaussian")
  )
```

#### Custom inference model

We define an overall F-test for linear regression as the inference model
and the p-value as the “significance” criterion:

``` r
# Define the ftest that returns a list of significance criteria
ftest <- function(y, x) {
  dat <- as.data.frame(cbind(y, x))
  lm_mod <- lm(y ~ ., data = dat)
  fstat <- summary.lm(lm_mod)$fstat
  fpval <- pf(fstat[1], fstat[2], fstat[3], lower.tail=F)
  names(fpval) <- "F-test"
  return(list(pval = fpval))
}
# define an InferenceModel using the F-test
imod <- mpower::InferenceModel(model = ftest, name = "F-test")
```

#### Run power analysis

We run power analysis for sample sizes 100 and 300 with 3 cores and 100
Monte Carlo simulations:

``` r
curve <- mpower::sim_curve(xmod, ymod_list, imod, s = 100, 
                   n = c(100, 300), cores = 3)
```

We can get a table of summaries and a plot of the power:

``` r
curve_df <- mpower::summary(curve , crit = "pval", thres = 0.05, how = "lesser")
```

<img src="man/figures/README-curve-summary-1.png" width="100%" />

    #> 
    #>  *** POWER CURVE ANALYSIS SUMMARY ***
    #> Number of Monte Carlo simulations: 100
    #> Number of observations in each simulation: 100 300
    #> Data generating process estimated SNR (for each outcome model): 0.26 0.06 0.01
    #> Inference model: F-test
    #> Significance criterion: pval
    #> Significance threshold: 0.05
    #> 
    #> |       | power|   n|  snr|
    #> |:------|-----:|---:|----:|
    #> |F-test |  0.96| 100| 0.26|
    #> |F-test |  0.54| 100| 0.06|
    #> |F-test |  0.08| 100| 0.01|
    #> |F-test |  1.00| 300| 0.26|
    #> |F-test |  0.95| 300| 0.06|
    #> |F-test |  0.27| 300| 0.01|

### Example 2: Estimate a data generative model from existing data

We can estimate a data generative model for the predictors if we have
existing data on them. For example, here we use a copy of the NHANES
data from the 2015-2016 and 2017-2018 cycles.

For an estimation model, all data must be in numeric form. Categorical
variables should be converted to factors with numeric categories:

``` r
data("nhanes1518")
# preprocessing step 
phthalates_names <- c("URXCNP", "URXCOP", "URXECP", "URXHIBP",
         "URXMBP", "URXMC1", "URXMCOH", "URXMEP", "URXMHBP", "URXMHH",
         "URXMHNC", "URXMHP", "URXMIB", "URXMNP", "URXMOH", "URXMZP")
covariates_names <- c("RIDAGEYR","INDHHIN2","BMXBMI","RIAGENDR")
# Get set of complete data only. Remaining 4800 obs.
nhanes_nona <- nhanes1518 %>%
  select(all_of(c(phthalates_names, covariates_names))) %>%
  filter(complete.cases(.))
```

#### Estimate the generative model:

We use a Bayesian semi-parametric Gaussian copula for this step. See
more in the R package `sbgcop`.

``` r
xmod <- mpower::MixtureModel(method = "estimation" , data = nhanes_nona, 
    sbg_args = list(nsamp = 1000))
```

``` r
mplot(xmod, split=F)$corr
```

<img src="man/figures/README-est-mplot-1.png" width="100%" />

#### Define an outcome model:

``` r
ymod <- mpower::OutcomeModel(sigma = 1, family = "gaussian", 
                     f = "0.5*RIDAGEYR + 0.3*I(RIAGENDR==1) +
                     0.3*URXMHH + 0.2*URXMOH + 0.2*URXMHP")
```

#### Estimate power using the F-test we define earlier:

``` r
power <- mpower::sim_power(xmod, ymod, imod, s=1000, n=100, snr_iter=4800,
                   cores = 3, errorhandling = "stop")
```

``` r
power_df <- mpower::summary(power , crit = "pval", thres = 0.05, how = "lesser")
#> 
#>  *** POWER ANALYSIS SUMMARY ***
#> Number of Monte Carlo simulations: 1000
#> Number of observations in each simulation: 100
#> Data generating process estimated SNR: 233.46
#> Inference model: F-test
#> Significance criterion: pval
#> 
#> Significance threshold:  0.05
#> 
#> |       | power|
#> |:------|-----:|
#> |F-test |     1|
```

### Example 3: Use the built-in inference models

#### Resample predictors from NHANES:

In this example, we generate predictors by resampling of a large
existing dataset. We load data from the R package NHANES and treat
ordinal Education and HHIncome as continuous variables.

``` r
library(NHANES)
data("NHANES")
nhanes_demo <- NHANES %>%
  select(BMI, Education, HHIncome, HomeOwn, Poverty, Age) %>%
  filter(complete.cases(.)) %>%
  mutate(Education = case_when(Education == "8th Grade" ~ 1,
                               Education == "9 - 11th Grade" ~ 2,
                               Education == "High School" ~ 3,
                               Education == "Some College" ~ 4,
                               Education == "College Grad" ~ 5),
         HHIncome = case_when(HHIncome == "more 99999" ~ 12,
                              HHIncome == "75000-99999" ~ 11,
                              HHIncome == "65000-74999" ~ 10,
                              HHIncome == "55000-64999" ~ 9,
                              HHIncome == "45000-54999" ~ 8,
                              HHIncome == "35000-44999" ~ 7,
                              HHIncome == "25000-34999" ~ 6,
                              HHIncome == "20000-24999" ~ 5,
                              HHIncome == "15000-19999" ~ 4,
                              HHIncome == "10000-14999" ~ 3,
                              HHIncome == " 5000-9999" ~ 2,
                              HHIncome == " 0-4999" ~ 1))
xmod <- mpower::MixtureModel(method = "resampling", data = nhanes_demo %>% select(- BMI))
```

#### Define the outcome model with a fitted model

We can generate outcomes from a previously fitted model.

``` r
lm_demo <- lm(BMI ~ Poverty*(poly(Age, 2) + HHIncome + HomeOwn + Education), data = nhanes_demo)
# wrapper function that takes a data frame x and returns a vector of outcome
f_demo <- function(x) {
  predict.lm(lm_demo, newdata = x)
}
ymod <- mpower::OutcomeModel(f = f_demo, family = "gaussian", 
    sigma = sigma(lm_demo))
```

#### Use the built-in `glm` model

We need to give `glm` the family and formula inputs. See documentation
of `glm` for more options.

``` r
# pass the formula argument to glm()
imod <- mpower::InferenceModel(model = "glm", family = "gaussian", 
    formula = y ~ Poverty*(poly(Age, 2) + HHIncome + HomeOwn + Education))
```

The model `lm_demo` is defined in the global environment, outside the
scope of the parallel processes. We need to pass `"lm_demo"` to the
`cluster_export` argument to make sure the variable is defined while
using parallelism.

``` r
curve <- mpower::sim_curve(xmod, ymod, imod, 
    s = 200, n = seq(1000, 5000, 1000), 
    cores = 4, errorhandling = "remove", snr_iter = 5000,
    cluster_export = c("lm_demo"))
```

``` r
curve_df <- mpower::summary(curve, crit = "pval", thres = 0.05, how = "lesser")
```

<img src="man/figures/README-resamp-summary-1.png" width="100%" />

    #> 
    #>  *** POWER CURVE ANALYSIS SUMMARY ***
    #> Number of Monte Carlo simulations: 200
    #> Number of observations in each simulation: 1000 2000 3000 4000 5000
    #> Data generating process estimated SNR (for each outcome model): 0.03
    #> Inference model: glm
    #> Significance criterion: pval
    #> Significance threshold: 0.05
    #> 
    #> |          | power|    n|  snr|
    #> |:---------|-----:|----:|----:|
    #> |Education | 0.330| 1000| 0.03|
    #> |HHIncome  | 0.185| 1000| 0.03|
    #> |HomeOwn   | 0.155| 1000| 0.03|
    #> |Poverty   | 0.575| 1000| 0.03|
    #> |Age       | 0.730| 1000| 0.03|
    #> |Education | 0.590| 2000| 0.03|
    #> |HHIncome  | 0.410| 2000| 0.03|
    #> |HomeOwn   | 0.220| 2000| 0.03|
    #> |Poverty   | 0.830| 2000| 0.03|
    #> |Age       | 0.955| 2000| 0.03|
    #> |Education | 0.700| 3000| 0.03|
    #> |HHIncome  | 0.470| 3000| 0.03|
    #> |HomeOwn   | 0.210| 3000| 0.03|
    #> |Poverty   | 0.880| 3000| 0.03|
    #> |Age       | 0.990| 3000| 0.03|
    #> |Education | 0.860| 4000| 0.03|
    #> |HHIncome  | 0.580| 4000| 0.03|
    #> |HomeOwn   | 0.205| 4000| 0.03|
    #> |Poverty   | 0.955| 4000| 0.03|
    #> |Age       | 1.000| 4000| 0.03|
    #> |Education | 0.890| 5000| 0.03|
    #> |HHIncome  | 0.720| 5000| 0.03|
    #> |HomeOwn   | 0.285| 5000| 0.03|
    #> |Poverty   | 0.970| 5000| 0.03|
    #> |Age       | 1.000| 5000| 0.03|

#### Usage of other built-in inference models

Here are examples of how to use other statistical inference models
included in the package.

###### Bayesian weighted sums (BWS)

“Significant” is when the credible intervals don’t include zero. We
access the 95% credible interval coverage with criterion “beta” and
threshold 0.05.

``` r
bws_imod <- mpower::InferenceModel(model = "bws", iter = 2000, family = "gaussian", refresh = 0)
bws_power <- mpower::sim_power(xmod, ymod, bws_imod, s = 100, n = 1000, 
                       cores=3, snr_iter=1000, errorhandling = "stop",
                       cluster_export = c("lm_demo"))
```

###### Quantile G-computation

Similar to BWS but is a frequentist method. The significance criterion
is the p-value.

``` r
qgcomp_imod <- mpower::InferenceModel(model="qgc")
qg_power <- mpower::sim_power(xmod, ymod, qgcomp_imod, s = 100, n = 1000, 
                       cores=3, snr_iter=1000, errorhandling = "remove")
```

###### Bayesian factor analysis with interactions

``` r
fin_imod <- InferenceModel(model="fin", nrun = 2000, verbose=F)
fin_power <- sim_power(xmod, ymod, fin_imod, s = 100, n = 1000, 
                       cores=3, snr_iter=1000, errorhandling = "remove")
```

###### Bayesian model averaging

``` r
bma_imod <- InferenceModel(model="bma", glm.family = "gaussian")
bma_power <- sim_power(xmod, ymod, bma_imod, s = 100, n = 1000, 
                       cores=3, snr_iter=1000, errorhandling = "remove")
```

###### Bayesian kernel machine regression

``` r
bkmr_imod <- InferenceModel(model = "bkmr", iter = 5000, verbose = F)
bkmr_power <- sim_power(xmod, ymod, bkmr_imod, s = 100, n = 1000, 
                       cores=3, snr_iter=1000, errorhandling = "remove")
```

###### MixSelect

``` r
ms_imod <- InferenceModel(model = "ms", nrun = 5000, verbose=F)
# Can expect run time for sample size of 1000 with 100 MC simulations 
# to be very long.
ms_power <- sim_power(xmod, ymod, ms_imod, s = 100, n = 1000, 
                       cores=1, snr_iter=1000, errorhandling = "stop")
```

### Example 4: Logistic regression example

We show how to work with a binary outcome here. We load NHANES data from
the 2015-2016 and 2017-2018 cycles included in our package and define a
“true” outcome model that includes linear main effects and an
interaction. We check the power to detect these effects using a logistic
regression and a sample size of 2000 observations.

``` r
data("nhanes1518")
chems <- c("URXCNP", "URXCOP", "URXECP", "URXHIBP", "URXMBP", "URXMC1", "URXMCOH", "URXMEP",
"URXMHBP", "URXMHH", "URXMHNC", "URXMHP", "URXMIB", "URXMNP", "URXMOH", "URXMZP")
chems_mod <- mpower::MixtureModel(nhanes1518[, chems] %>% filter(complete.cases(.)),
                                  method = "resampling")
bmi_mod <- mpower::OutcomeModel(f = "0.2*URXCNP + 0.15*URXECP + 0.1*URXCOP*URXECP", family = "binomial")
logit_mod <- mpower::InferenceModel(model = "glm", family = "binomial")
logit_out <- mpower::sim_power(xmod=chems_mod, ymod=bmi_mod, imod=logit_mod,
                 s=100, n=2000, cores=2, snr_iter=5000)
```

``` r
logit_df <- summary(logit_out, crit="pval", thres=0.05, how="lesser")
#> 
#>  *** POWER ANALYSIS SUMMARY ***
#> Number of Monte Carlo simulations: 100
#> Number of observations in each simulation: 2000
#> Data generating process estimated SNR: 0.65
#> Inference model: glm
#> Significance criterion: pval
#> 
#> Significance threshold:  0.05
#> 
#> |        | power|
#> |:-------|-----:|
#> |URXCNP  |  0.17|
#> |URXCOP  |  0.99|
#> |URXECP  |  0.93|
#> |URXHIBP |  0.03|
#> |URXMBP  |  0.02|
#> |URXMC1  |  0.03|
#> |URXMCOH |  0.02|
#> |URXMEP  |  0.04|
#> |URXMHBP |  0.03|
#> |URXMHH  |  0.04|
#> |URXMHNC |  0.04|
#> |URXMHP  |  0.07|
#> |URXMIB  |  0.03|
#> |URXMNP  |  0.05|
#> |URXMOH  |  0.08|
#> |URXMZP  |  0.03|
```


NEWS.md
# mpower 0.1.0

First version to go to CRAN.


R/sim.R
#' Power curve using Monte Carlo simulation
#'
#' This function can be used to create power curves by calling sim_power() on
#' combinations of many sample sizes and signal-to-noise ratio (SNR).
#'
#' @param xmod A MixtureModel object.
#' @param ymod One or a list of OutcomeModel object(s).
#' @param imod An InferenceModel object.
#' @param s An integer for the number of Monte Carlo simulations.
#' @param n An integer or a vector of sample sizes.
#' @param cores An integer for the number of processing cores. When cores > 1,
#'   parallelism is automatically applied.
#' @param file A string, a file name with no extension to write samples to
#'   periodically. By default write to an RDS file.
#' @param errorhandling A string "remove", "stop", or "pass". If an error occurs
#'   in any iteration, remove that iteration (remove), return the error message
#'   verbatim in the output (pass), or terminate the loop (stop). Default is
#'   "remove". See R package 'foreach' for more details.
#' @param snr_iter An integer for number of Monte Carlo samples to estimate SNR.
#' @param cluster_export A vector of functions to pass to the
#'   parallel-processing clusters.
#' @return A SimCurve object with the following attributes:
#' \describe{
#'   \item{s}{a number of simulations.}
#'   \item{snr}{a real number or array of real numbers for SNR of each OutcomeModel.}
#'   \item{n}{a number or vector of sample sizes.}
#'   \item{xmod}{the MixtureModel used.}
#'   \item{ymod}{the OutcomeModel used.}
#'   \item{imod}{the InferenceModel used.}
#'   \item{sims}{a list of simulation output matrices.}
#' }
#' @examples
#' data("nhanes1518")
#' chems <- c("URXCNP", "URXCOP", "URXECP", "URXHIBP", "URXMBP", "URXMC1",
#' "URXMCOH", "URXMEP","URXMHBP", "URXMHH", "URXMHNC", "URXMHP", "URXMIB",
#' "URXMNP", "URXMOH", "URXMZP")
#' chems_mod <- mpower::MixtureModel(nhanes1518[, chems], method = "resampling")
#' bmi_mod <- mpower::OutcomeModel(f = "0.2*URXCNP + 0.15*URXECP +
#' 0.1*URXCOP*URXECP", family = "binomial")
#' logit_mod <- mpower::InferenceModel(model = "glm", family = "binomial")
#' logit_out <- mpower::sim_curve(xmod=chems_mod, ymod=bmi_mod, imod=logit_mod,
#' s=50, n=c(500, 1000), cores=1, snr_iter=1000)
#' logit_df <- summary(logit_out, crit="pval", thres=0.05, how="lesser")
#' @export
sim_curve <- function(xmod,
                      ymod,
                      imod,
                      s = 100,
                      n = 100,
                      cores = 1,
                      file = NULL,
                      errorhandling = "stop",
                      snr_iter = 10000,
                      cluster_export = c()
                      ) {
  stopifnot(inherits(xmod, "mpower_estimation_MixtureModel") |
              inherits(xmod, "mpower_cvine_MixtureModel") |
              inherits(xmod, "mpower_resampling_MixtureModel"))
  stopifnot(inherits(imod, "mpower_InferenceModel"))
  if (inherits(ymod, "mpower_OutcomeModel")) {
    ymod <- list(ymod)
  }
  sim_list <- list()
  ymod_list <- list()
  n_vec <- c()
  snr_vec <- rep(0, length(ymod))
  cur_file <- NULL
  k <- 1
  for (i in seq_along(n)) {
    cur_n <- n[i]
    for (j in seq_along(ymod)) {
      cur_ymod <- ymod[[j]]
      cur_file <- if(!is.null(file)) {paste(file, i, j, sep = "-")}
      print(paste("Simulation for n =", cur_n, "and the", j, "th outcome model"))
      out <- sim_power(xmod = xmod,
                       ymod = cur_ymod,
                       imod = imod,
                       s = s,
                       n = cur_n,
                       snr_iter = snr_iter,
                       cores = cores,
                       file = cur_file,
                       errorhandling = errorhandling,
                       cluster_export = cluster_export)
      sim_list[[k]] <- out$sims
      ymod_list[[k]] <- out$ymod
      snr_vec[j] <- snr_vec[j] + out$snr
      n_vec <- c(n_vec, out$n)
      k <- k+1
    }
  }
  new_SimCurve(list(s = s, n = n_vec,
                    snr = rep(snr_vec/length(n), length(n)),
                    xmod = xmod,
                    ymod = ymod_list,
                    imod = imod,
                    sims = sim_list))
}

new_SimCurve <- function(x = list()) {
  stopifnot(is.list(x))
  structure(x, class = "mpower_SimCurve")
}

#' Power analysis using Monte Carlo simulation
#'
#' @param xmod A MixtureModel object.
#' @param ymod An OutcomeModel object.
#' @param imod An InferenceModel object.
#' @param s An integer for number of Monte Carlo simulations.
#' @param n An integer for sample size in each simulation.
#' @param cores An integer for number of processing cores. When cores > 1,
#'   parallelism is automatically applied.
#' @param file A string, a file name with no extension to write samples to
#'   periodically. By default write to an RDS file.
#' @param errorhandling A string "remove", "stop", or "pass". If an error occurs
#'   in any iteration, remove that iteration (remove), return the error message
#'   verbatim in the output (pass), or terminate the loop (stop). Default is
#'   "remove". See R package `foreach` for more details.
#' @param snr_iter An integer for number of Monte Carlo samples to estimate SNR
#' @param cluster_export A vector of functions to pass to the parallel-processing clusters
#' @return A PowerSim object. Attributes:
#' \describe{
#'   \item{s}{a number of simulations.}
#'   \item{snr}{a real number for SNR of the OutcomeModel.}
#'   \item{n}{a number of sample sizes.}
#'   \item{xmod}{the MixtureModel used.}
#'   \item{ymod}{the OutcomeModel used.}
#'   \item{imod}{the InferenceModel used.}
#'   \item{sims}{an output matrices.}
#' }
#' @examples
#' data("nhanes1518")
#' chems <- c("URXCNP", "URXCOP", "URXECP", "URXHIBP", "URXMBP", "URXMC1",
#' "URXMCOH", "URXMEP","URXMHBP", "URXMHH", "URXMHNC", "URXMHP", "URXMIB",
#' "URXMNP", "URXMOH", "URXMZP")
#' chems_mod <- mpower::MixtureModel(nhanes1518[, chems], method = "resampling")
#' bmi_mod <- mpower::OutcomeModel(f = "0.2*URXCNP + 0.15*URXECP +
#' 0.1*URXCOP*URXECP", family = "binomial")
#' logit_mod <- mpower::InferenceModel(model = "glm", family = "binomial")
#' logit_out <- mpower::sim_power(xmod=chems_mod, ymod=bmi_mod, imod=logit_mod,
#' s=100, n=2000, cores=1, snr_iter=2000)
#' logit_df <- summary(logit_out, crit="pval", thres=0.05, how="lesser")
#' @export
sim_power <- function(xmod,
                      ymod,
                      imod,
                      s = 100,
                      n = 100,
                      cores = 1,
                      file = NULL,
                      errorhandling = "stop",
                      snr_iter = 10000,
                      cluster_export = c()
                      ) {
  stopifnot(inherits(xmod, "mpower_estimation_MixtureModel") |
              inherits(xmod, "mpower_cvine_MixtureModel") |
              inherits(xmod, "mpower_resampling_MixtureModel"))
  stopifnot(inherits(ymod, "mpower_OutcomeModel"))
  stopifnot(inherits(imod, "mpower_InferenceModel"))
  pb <- utils::txtProgressBar(max = s, style = 3)
  progress <- function(i) utils::setTxtProgressBar(pb, i)
  if (cores < 2) {
    out <- as.list(rep(NA, s))
    for (i in seq_along(out)) {
      out[[i]] <- tryCatch({
        X <- genx(xmod, n)
        y <- geny(ymod, X)
        fit(imod, X, y)
        },
        error=function(e) {
          message(e)
          return(NA)
        })
      progress(i)
      if (!is.null(file) & (i%%50==0)) saveRDS(out, file = paste0(file, ".rds"))
    }
  } else {
    # Set up parallel backend to use many processors
    cores_max <- parallel::detectCores()
    cl <- snow::makeCluster(min(cores_max[1]-1, cores))
    snow::clusterExport(cl, c(c("genx", "geny", "fit"), cluster_export))
    doSNOW::registerDoSNOW(cl)
    opts <- list(progress=progress)
    out <- foreach::foreach(i = 1:s,
                            .errorhandling = errorhandling,
                            .options.snow = opts) %dopar% {
      X <- genx(xmod, n)
      y <- geny(ymod, X)
      r <- fit(imod, X, y)
      ## write results to disk
      if (!is.null(file)) saveRDS(r, file = paste0(file, i, ".rds"))
      return(r)
    }
    snow::stopCluster(cl)
  }
  close(pb)
  snr <- estimate_snr(ymod, xmod, m = snr_iter)$est
  new_Sim(list(s = s,
               n = n,
               snr = snr,
               xmod = xmod,
               ymod = ymod,
               imod = imod,
               sims = out))
}

new_Sim <- function(x = list()) {
  stopifnot(is.list(x))
  structure(x, class = "mpower_Sim")
}

#' Tabular and graphical summaries of power simulation
#'
#' @param sim A Sim or a SimCurve object, output from `sim_power()` or
#'   `sim_curve()`.
#' @param crit A string specifying the significance criteria.
#' @param thres A number or vector of numbers specifying the thresholds of
#'   "significance".
#' @param digits An integer for the number of decimal points to display.
#' @param how A string, whether to compare the criterion 'greater' or 'lesser'
#'   than the threshold.
#' @return A data.frame summary of power for each predictor for each combination
#'   of thresholds, sample size, signal-to-noise ratios.
#' @export
summary <- function(sim, crit, thres, digits = 3, how = "greater") {
  UseMethod("summary", sim)
}

#' @export
summary.mpower_Sim <- function(sim, crit, thres, digits = 3, how = "greater") {
  if(seq_along(crit) > 1)  stop("Summary limited to one criteria", call. = FALSE)
  if (length(thres) == 1) {
    # For one threshold and one Sim object
    res <- summary_one_sim_one_thres(sim$sims, crit, thres, digits, how, pivot = T)
  } else if (length(thres) > 1) {
    # For many threshold and one Sim object
    res <- summary_one_sim_many_thres(sim$sims, crit, thres, digits, how)
  }
  colnames(res) <- c("thres", " ", "power")
  cat("\n\t*** POWER ANALYSIS SUMMARY ***")
  cat("\nNumber of Monte Carlo simulations:", sim$s)
  cat("\nNumber of observations in each simulation:", sim$n)
  cat("\nData generating process estimated SNR:", round(sim$snr,2))
  cat("\nInference model:", sim$imod$model_name)
  cat("\nSignificance criterion:", crit)
  for (tt in thres) {
    cat("\n\nSignificance threshold: ", tt)
    cat("\n", res %>% dplyr::filter(thres == tt) %>%
          dplyr::select(-thres) %>% knitr::kable(), sep="\n")
  }
  return(res)
}

#' @export
summary.mpower_SimCurve <- function(sim, crit, thres, digits = 3, how = "greater") {
  if (length(thres) == 1) {
    # For one threshold and many Sim objects
    res <- summary_many_sim_one_thres(sim$sims, sim$n, sim$snr, crit, thres, digits, how)
  } else {
    warning("Using the first threshold only")
    res <- summary_many_sim_one_thres(sim$sims, sim$n, sim$snr, crit, thres[1], digits, how)
  }
  cat("\n\t*** POWER CURVE ANALYSIS SUMMARY ***")
  cat("\nNumber of Monte Carlo simulations:", sim$s)
  cat("\nNumber of observations in each simulation:", unique(sim$n))
  cat("\nData generating process estimated SNR (for each outcome model):",
      round(unique(sim$snr),2))
  cat("\nInference model:", sim$imod$model_name)
  cat("\nSignificance criterion:", crit)
  cat("\nSignificance threshold:", thres[1])
  cat("\n", res %>%
        dplyr::select(tidyselect::all_of(c("test", "power", "n", "snr"))) %>%
        magrittr::set_colnames(c(" ", "power", "n", "snr")) %>%
        knitr::kable(), sep="\n")
  return(res)
}

summary_many_sim_one_thres <- function(sim, n, snr, crit, thres, digits, how) {
  stopifnot(is.list(sim))
  num_sims <- length(sim)
  res <- tibble::tibble()
  for (i in 1:num_sims) {
    res <- dplyr::bind_rows(res,
                            summary_one_sim_one_thres(sim[[i]], crit, thres, digits, how, pivot = F))
  }
  res <- res %>%
    dplyr::mutate(n = n, snr = round(snr, 2)) %>%
    tidyr::pivot_longer(cols = -c("n", "snr", "thres"), names_to = "test", values_to = "power")
  if (length(unique(res$snr)) > 1) {
    g <- res %>%
      dplyr::mutate(snr = as.factor(snr)) %>%
      ggplot(aes(x = !! sym("n"), y = !! sym("power"), colour = !! sym("snr"))) +
      geom_hline(yintercept = 0.8, linetype = 2, colour = "grey") +
      geom_path() + geom_point() +
      labs(x="Sample size", y="Type I error rate/ Power") +
      scale_colour_discrete(name = "SNR") +
      facet_wrap(stats::as.formula("~test"))
  } else {
    g <- res %>%
      ggplot(aes(x = !! sym("n"), y = !! sym("power"), colour = !! sym("test"))) +
      geom_hline(yintercept = 0.8, linetype = 2, colour = "grey") +
      geom_path() + geom_point() +
      labs(x="Sample size", y="Type I error rate/ Power") +
      scale_colour_discrete(name = " ")
  }
  print(g)
  return(res)
}

summary_one_sim_many_thres <- function(sim, crit, thres, digits, how) {
  res <- data.frame()
  for (i in seq_along(thres)) {
    res <- dplyr::bind_rows(res,
                            summary_one_sim_one_thres(sim, crit, thres[i], digits, how, pivot = T))
  }
  g <- res %>%
    ggplot(aes(x = !! sym("thres"), y = !! sym("power"),
               group = !! sym("test"), color = !! sym("test"))) +
    geom_hline(yintercept = 0.8, linetype = 2, colour = "gray") +
    geom_point() + geom_path() +
    labs(y = "Type I error rate/ Power", x = "Significance threshold") +
    scale_colour_discrete(name = "")
  print(g)
  return(res)
}

# Returns a data.frame. Each column is power of a predictor. A column for threshold.
summary_one_sim_one_thres <- function(sim, crit, thres, digits, how, pivot=FALSE) {
  res <- sim %>%
    purrr::map(crit) %>%
    purrr::map(function(e) {
      if (how == "greater") {e >= thres}
      else {e <= thres}}) %>%
    abind::abind(along = -1) %>%
    tibble::as_tibble() %>%
    dplyr::summarise(dplyr::across(tidyselect::everything(), mean)) %>%
    round(digits) %>%
    dplyr::mutate(thres = thres)
  if (pivot) {
    res <- res %>%
      tidyr::pivot_longer(-thres, names_to = "test", values_to = "power")
  }
  return(res)
}
